luthier/src/test/groovy/com/yahoo/bard/webservice/config/luthier/factories/KeyValueStoreDimensionFactorySpec.groovy

```diff
@@ -0,0 +1,70 @@
+import com.yahoo.bard.webservice.data.dimension.Dimension
+import com.yahoo.bard.webservice.data.dimension.DimensionField
+import com.yahoo.bard.webservice.data.dimension.MapStore
+import com.yahoo.bard.webservice.data.dimension.SearchProvider
+import com.yahoo.bard.webservice.data.dimension.impl.NoOpSearchProvider
+import spock.lang.Specification
+
+class KeyValueStoreDimensionFactorySpec extends Specification {
+    LuthierIndustrialPark park
+    Dimension testDimension
+    void setup() {
+        Map<String, Factory<Dimension>> dimensionFactoriesMap = new HashMap<>()
+        dimensionFactoriesMap.put("KeyValueStoreDimensionFactory", new KeyValueStoreDimensionFactory())
+
+        ResourceDictionaries resourceDictionaries = new ResourceDictionaries()
+        park = new LuthierIndustrialPark.Builder(resourceDictionaries)
+                .withDimensionFactories(dimensionFactoriesMap)
+                .build()
+        park.load()
+        testDimension = park.getDimension("testDimension")
+    }
+
+    //TODO: .is check chaching
+
+    def "The spec content of a specific test dimension matches with the JSON config file and defaults"() {
+        when:
+            // set up for: fields content correctness
+            LinkedHashSet<DimensionField> dimensionFields = testDimension.getDimensionFields()
+            LuthierDimensionField key = testDimension.getKey()
+
+            SearchProvider searchProvider = testDimension.getSearchProvider()
+        then:
+            // basic String correctness
+            testDimension.getApiName() == "testDimension"
+            testDimension.getLongName() == "a longName for testing"
+            testDimension.getCategory() == "UNKNOWN_CATEGORY"
+            testDimension.getDescription() == "a description for testing"
+            testDimension.getStorageStrategy().getApiName() == "loaded"
+            testDimension.isAggregatable() == true
+
+            // Fields content correctness
+            List expectedCamelNames = ["testPk", "testField1", "testField2", "testField3"]
+            List expectedFieldTags = [ ["primaryKey"], [], [], [] ]
+            // next line is to be uncommented after we implemented the description correctly
+            // List expectedFieldDescriptions = [ "", "", "", "" ]
+            for (int i = 0; i < dimensionFields.size(); i++) {
+                assert dimensionFields[i].getName() == expectedCamelNames[i]
+                assert dimensionFields[i].getTags() == expectedFieldTags[i]
+                assert dimensionFields[i].getDescription() == "Error: currently there is no description"
+            }
+            key == dimensionFields[0]
+
+            // type correctness of searchProvider
+            // we cannnot check keyValueStore directly
+            searchProvider.getClass() == NoOpSearchProvider
+    }
+
+    def "When new dimension's name match with the previous dimension, use a reference to the previous one"() {
```

Comment:
This test shouldn't be passing. The fact that it does means there is some
state leaking between tests (i.e the resource dictionary cache). This is bad.
We don't want state leaking between tests, because it imposes an implicit
ordering on our tests. An implicit ordering on tests can lead to sporadic test
failures. For example, this test will fail if it's executed before the test
above it, and there is *no guarantee* on the order the test framework executes
tests. So we need to make a few changes:

1. We need to add a cleanup method that clears the resource dictionaries
between tests.

2. We need to update this test so that it queries industrial park for the same
dimension twice. That way it's standalone.

Reply:
Disregard this comment. I missed the setup method when I was reviewing this test, and incorrectly thought the `testDimension` was being initialized in the first test.